jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CrossOriginHandler.java

```diff
@@ -0,0 +1,423 @@
+ * the origin server with origin {@code http://domain.com}.</p>
+ */
+@ManagedObject
+public class CrossOriginHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(CrossOriginHandler.class);
+
+    private boolean allowCredentials = true;
+    private Set<String> allowedHeaders = Set.of("Content-Type");
+    private Set<String> allowedMethods = Set.of("GET", "POST", "HEAD");
+    private Set<String> allowedOrigins = Set.of("*");
+    private Set<String> allowedTimingOrigins = Set.of();
+    private boolean deliverPreflight = false;
+    private Set<String> exposedHeaders = Set.of();
+    private Duration preflightMaxAge = Duration.ofSeconds(60);
+    private boolean anyOriginAllowed;
+    private final Set<Pattern> allowedOriginPatterns = new LinkedHashSet<>();
+    private boolean anyTimingOriginAllowed;
+    private final Set<Pattern> allowedTimingOriginPatterns = new LinkedHashSet<>();
+
+    /**
+     * @return whether the cross server allows cross-origin requests to include credentials
+     */
+    @ManagedAttribute("Whether the server allows cross-origin requests to include credentials (cookies, authentication headers, etc.)")
+    public boolean isAllowCredentials()
+    {
+        return allowCredentials;
+    }
+
+    /**
+     * <p>Sets whether the cross server allows cross-origin requests to include credentials
+     * such as cookies or authentication headers.</p>
+     * <p>For example, when the cross server allows credentials to be included, cross-origin
+     * requests will contain cookies, otherwise they will not.</p>
+     * <p>The default is {@code true}.</p>
+     *
+     * @param allow whether the cross server allows cross-origin requests to include credentials
+     */
+    public void setAllowCredentials(boolean allow)
+    {
+        throwIfStarted();
+        allowCredentials = allow;
+    }
+
+    /**
+     * @return the set of allowed headers in a cross-origin request
+     */
+    @ManagedAttribute("The set of allowed headers in a cross-origin request")
+    public Set<String> getAllowedHeaders()
+    {
+        return allowedHeaders;
+    }
+
+    /**
+     * <p>Sets the set of allowed headers in a cross-origin request.</p>
+     * <p>The cross server receives a preflight request that specifies the headers
+     * of the cross-origin request, and the cross server replies to the preflight
+     * request with the set of allowed headers.
+     * Browsers are responsible to check whether the headers of the cross-origin
+     * request are allowed, and if they are not produce an error.</p>
+     * <p>The headers can be either the character {@code *} to indicate any
+     * header, or actual header names.</p>
+     *
+     * @param headers the set of allowed headers in a cross-origin request
+     */
+    public void setAllowedHeaders(Set<String> headers)
+    {
+        throwIfStarted();
+        allowedHeaders = headers;
```

Comment:
Strange that this is pass by reference, when most of our configuration is done by pass by value.
Somebody could pass in an immutable set, and then later configuration break when somebody tries to mutate the return from `getAllowedHeaders()`.

Reply:
Okay now I am calling `Set.copyOf()` on the setters that take a `Set`.